flink-ml-lib/src/main/java/org/apache/flink/ml/common/broadcast/operator/AbstractBroadcastWrapperOperator.java

```diff
@@ -232,6 +247,78 @@ private OperatorMetricGroup createOperatorMetricGroup(
         }
     }
 
+    /**
+     * extracts common processing logic in subclasses' processing elements.
+     *
+     * @param streamRecord the input record.
+     * @param inputIndex input id, starts from zero.
+     * @param consumer the consumer function.
+     * @throws Exception
+     */
+    protected void processElementX(
+            StreamRecord streamRecord,
+            int inputIndex,
+            ThrowingConsumer<StreamRecord, Exception> consumer)
+            throws Exception {
+        if (!isBlocked[inputIndex]) {
+            if (areBroadcastVariablesReady()) {
+                dataCacheWriters[inputIndex].finishCurrentSegmentAndStartNewSegment();
```

Comment:
Perhaps we also extract the method to process pending records to eliminate the repeat with `endInputX`?

Reply:
Thanks @gaoyunhaii ! I have extracted `AbstractBroadcastWrapperOperator#processPendingElements`.